pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -381,3 +390,45 @@ def register_plugins(cls, plugin_registry: Dict) -> None:
             description="DDP Plugin with `find_unused_parameters` as False",
             find_unused_parameters=False
         )
+
+    def _share_information_to_prevent_deadlock(self):
+        self._share_pids()
+
+        # remove `PL_DDP_SYNC_TMPDIR` from os.environ
+        self._sync_dir = os.environ.pop("PL_DDP_SYNC_TMPDIR", None)
+
+    def _share_pids(self):
+        """
+        Make all DDP processes aware of all processes pids.
+        """
+        self.barrier()
+        pids = self.all_gather(torch.tensor(os.getpid(), device=self.root_device))
+        pids = pids.cpu().numpy().tolist()
+        self._pids = pids if isinstance(pids, list) else [pids]
+
+    def reconciliate_processes(self, trace: str):
+        if self.world_size > 1:
+            sync_dir = self._sync_dir
+
+            if not os.path.exists(sync_dir):
+                # avoid race condition
+                try:
+                    os.makedirs(sync_dir)
+                except FileExistsError:
+                    pass
+
+            # save a file locally.
+            torch.save(trace, os.path.join(sync_dir, f"{self.global_rank}.pl"))
+
+            # sleep for a short time
+            time.sleep(3)
+
+            # return if all processes wrote a file in the `sync_dir`.
+            if len(os.listdir(sync_dir)) == self.world_size:
+                return
```

Comment:
this assumes there is a shared filesystem, which is the case most of the time. 
in the other case, it will mistakenly detect a deadlock. 
can you write it in the docstring that we expect this to fail, so at least it is documented somewhere?

Reply:
Done !